src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1350,12 +1484,48 @@ DatabaseShardImp::getCache(std::uint32_t seq)
     return std::make_pair(pCache, nCache);
 }
 
-std::uint64_t
-DatabaseShardImp::available() const
+bool
+DatabaseShardImp::sufficientStorage(
+    std::uint32_t numShards,
+    bool isHistorical,
+    std::lock_guard<std::mutex> const&) const
 {
     try
     {
-        return boost::filesystem::space(dir_).available;
+        std::vector<std::uint64_t> capacities;
+
+        // Get the available storage for each historical path
+        if (isHistorical && !historicalPaths_.empty())
+        {
+            std::for_each(
+                historicalPaths_.begin(),
+                historicalPaths_.end(),
+                [&capacities](boost::filesystem::path const& path) {
+                    capacities.push_back(
+                        boost::filesystem::space(path).available);
+                });
+        }
+
+        // Get the available storage for the main shard path
+        else
+            capacities.push_back(boost::filesystem::space(dir_).available);
+
+        return std::find_if(
+                   capacities.begin(),
+                   capacities.end(),
+
+                   // Leverage all the historical shard paths to
+                   // see if collectively they can fit the specified
+                   // number of shards. For this to work properly,
+                   // each historical path must correspond to a separate
+                   // physical device.
+                   [&numShards, this](std::uint64_t const& capacity) {
+                       numShards -= std::min(
+                           capacity / avgShardFileSz_,
+                           std::uint64_t(numShards));
+
+                       return numShards == 0;
+                   }) != capacities.end();
```

Comment:
Since you're modifying this file, Line 1534 (not part of this changeset) should be return false not return 0.

Reply:
Done